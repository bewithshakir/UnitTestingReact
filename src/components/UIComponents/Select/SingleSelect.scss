.select{
    width:100%;

    & .MuiInputLabel-outlined.MuiInputLabel-shrink {
        transform: translate(2px, 0px) scale(0.75);
        z-index: 0;
    }

    .react-select-container{
        &.error {
        border: 1px solid var(--Tertiary);
        box-shadow: none;
            &  .react-select{
                &__control{
                border: none;
                box-shadow: none;
                    &:hover, &:active, &--is-focused{
                        border: none;
                        box-shadow: none;
                    }
                    &--menu-is-open{
                    & .react-select__indicator{
                        padding-left: 18px;
                        transform: rotate(180deg);
                    }
                }
                }
            }
        }    

        & .react-select{
            &__control{
                height: 50px;
                border-radius: 0px;
                font-size: 14px;
                &--menu-is-open{
                    & .react-select__indicator{
                        padding-left: 18px;
                        transform: rotate(180deg);
                    }
                }
                &:hover, &:active, &--is-focused{
                    height: 50px;
                    border-radius: 0;
                    border-color: #404040;
                    border: 1px solid var(--Darkgray);
                    box-shadow: none;
                }
            }
            &__value-container{
                padding: 2px 10px;
                &__placeholder{
                    font-size: 14px;
                    line-height: 16px;
                }
            }
            &__indicator{
                padding-right: 18px;
            }
        }

    }

    .react-select{
        &__menu{
            border-radius: 0;
            background:var(--White);
            box-shadow: 3px 6px 14px #00000026;
            z-index: 999;
        }
        &__menu-list{
            max-height: 256px;
            overflow-y: scroll;
        }    
        &__option {
            height:50px;
            line-height: 27px;
            font-size: 14px;  
            padding: 10px 15px;
            box-sizing: border-box;
            border-bottom: 1px solid var(--White);
            cursor: pointer;

                &--is-focused{   
                background-color: var(--Lightgray_2);
                opacity: 1;
                    &:active {
                        border: 1px solid var(--White);
                        background-color: var(--Gray);
                        color: var(--White);
                    } 
                }
                &--is-selected{
                    border: 1px solid var(--White);
                    background-color: var(--Gray);
                    color: var(--White);
                } 
            }
        &__menu-portal {
            background:var(--White);
            box-shadow: 3px 6px 14px #00000026;
            z-index: 999;
        }    
     }

}
