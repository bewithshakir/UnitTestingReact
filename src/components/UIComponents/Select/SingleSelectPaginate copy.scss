.select {
  width: 100%;
}

div{
  .select-paginate-custom{
    .css-qc6sy-singleValue, .css-14el2xx-placeholder, .css-319lph-ValueContainer, .css-1s2u09g-control, .css-14dclt2-Input, .css-1pahdxg-control{
      height: 50px;
      border-radius: 0px;
      font-size: 14px;
      padding: 0px;
      margin: 0px;
      cursor:text;
      outline: none;
    }

    .css-1hb7zxy-IndicatorsContainer{
      display: none;
    }

    .css-qc6sy-singleValue{
      padding: 14px 10px 15px;
    }

    .css-14dclt2-Input{
      margin: 0px;
      padding: 10px 10px 15px;

    }

    .css-14el2xx-placeholder{
      padding: 14px 10px 15px;
    }

    .css-26l3qy-menu{
      border-radius: 0;
      padding: 0px;
      // margin: 0px;
    }

    .css-4ljt47-MenuList{
      border-radius: 0;
      background: var(--White);
      box-shadow: 3px 6px 14px #00000026;
      padding: 0px;
      z-index: 999;
      .css-9gakcf-option, .css-yt9ioa-option, .css-1n7v3ny-option{
        height: 50px;
        line-height: 27px;
        font-size: 14px;
        padding: 10px 15px;
        box-sizing: border-box;
        border-bottom: 1px solid var(--White);
        cursor: pointer;
        display: flex;
      }
      // .css-yt9ioa-option{
      //   background-color: var(--Lightgray_2);
      //   opacity: 1;
      //   font-size: 14px;
      // }
      .css-1n7v3ny-option{
        background-color: var(--Lightgray_2);
        opacity: 1;
        font-size: 14px;
      }
      .css-9gakcf-option{
        border: 1px solid var(--White);
        background-color: var(--Gray);
        color: var(--White);
      }

    }


  }
} 