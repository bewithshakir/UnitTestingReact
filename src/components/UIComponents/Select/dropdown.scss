svg {
  display: block;
  width: 1.5em;
  height: 1.5em;
  fill: var(--Lightgray);
}

.select {
  font-family: arial;
  font-size: 14px;
  position: relative;
  display: inline-block;
  box-sizing: border-box;
  
  &:focus {
    outline: 0;
    
    & .selection {
      box-shadow: 0 0 1px 1px var(--Gray);
    }
  }
}

.label {
  color: var(--Darkgray) !important;
  font-size: 14px;
  font-family: arial;
  display: block;
  margin-bottom: 6px;
  font-weight: 700;
}

.selection {
  position: relative;
  padding: 5px;
  height: 50px;
  box-sizing: border-box;
  border: 1px solid var(--Lightgray_2);
  background: var(--White) 0% 0% no-repeat padding-box;
  &:focus {
        background: var(--White) 0% 0% no-repeat padding-box;
        color: var(--Darkgray);  
        border:2px solid var(--Darkgray);
      }
  &:active {
        background: var(--White) 0% 0% no-repeat padding-box;
        color: var(--Darkgray); 
        border: 1px solid var(--Gray);
      }     
}

.open{
      background: var(--White) 0% 0% no-repeat padding-box;
      color: var(--Darkgray);  
      border:2px solid var(--Darkgray);
}

.value {
  position: relative;
  display: inline-block;
  padding: 5px 10px;
  margin-top: 5px;
}

.multiple {
  padding-right: 30px;
  margin-right: 5px;
  background: var(--White);
  color: var(--Gray);
}

.delete {
  position: absolute;
  top: 0;
  right: 0;
  display: block;
  padding: 10px;
  font-size: 10px;
  cursor: pointer;
}

.placeholder {
  padding: 5px 10px;
  color: var(--Gray);
  margin-top: 5px;
}

.arrow {
  position: absolute;
  top: 5px;
  right: 5px;
  display: block;
  padding: 10px;
  font-size: 10px;
  color: var(--Gray);
}

.options {
  position: absolute;
  top: 115%;
  width:115%;
  left: 0;
  right: 0;
  border: solid var(--Lightgray_4);
  border-width: 0 1px;
  background:var(--White);
  box-shadow: 3px 6px 14px #00000026;
  z-index: 999;
  box-sizing: border-box;
  padding: 0 10px;
}

.opt-items{
  display: flex;
  height: 200px;
  overflow-y: scroll;
  flex-direction: column;
  width: 100%;
  overflow-x: auto;
}

.opt-search {
  display: flex;
  flex-direction: column;
}

.option {
  line-height: 1.5rem;
  font-size: 14px;  
  padding: 10px 15px;
  border-bottom: 1px solid var(--White);
  cursor: pointer;

  &.selected {
    border: 1px solid var(--White);
    margin: -1px -1px 0;
    background: var(--Lightgray_3);
  }
  
  &.focused {
    background: var(--Lightgray_3);
    opacity: 1;
  }
}

.checkbox {
  content: '';
  vertical-align: top;
  display: inline-block;
  width: 16px;
  height: 16px;
  padding: 2px;
  border: 1px solid var(--Darkgray);
  border-radius: 2px;
  margin: 2px 12px 0 0;
  color: var(--White);
  font-size: 10px;
  
  .selected & {
    border-color:var(--Black);
    background: var(--White);
  }
}

.ui.search .input{
    width :100%;
    height: 50px;
    box-sizing: border-box;
    display: flex;
    font-size: 14px;
    align-items: center;
    margin: 0;
    border:0;
}

.ui.search .prompt{
    height: 90%;
    width: 97%; 
    border-radius: 0;
    font-size: 14px;
}

.ui.input > input {
    border: 1px solid var(--White);
    font-size: 14px;
}

.icon{
    float: right;
    cursor: pointer;
    padding-left: 12px;
}

.ui.search input{
    &:focus-visible{
        border: 1px solid var(--White);
        outline-color: var(--White);
    }
}

.divider{
  height: 1px;
  margin: 5px 0;
  background-color: var(--White);
  border-bottom: 1px solid var(--Lightgray_2);
}