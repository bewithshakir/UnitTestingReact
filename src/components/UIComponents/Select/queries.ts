
import { useQuery } from "react-query";
import { AxiosRequestConfig } from "axios";
import axios from "../../../infrastructure/ApiHelper";

interface Payload {
    [k: string]: any
}

// let cancelTokenSoruce: CancelTokenSource | null;

export const getDataFromServer = async (url: string, payload: Payload) => {
    // if (cancelTokenSoruce) {
    //     cancelTokenSoruce.cancel();
    // }
    // cancelTokenSoruce = CancelToken.source();

    const query = new URLSearchParams();
    for (const [key, value] of Object.entries(payload)) {
        query.append(key, value);
    }
    const queryString = query.toString();
    const options: AxiosRequestConfig = {
        method: 'get',
        url: url + (queryString ? `?${queryString}` : ""),
        // cancelToken: cancelTokenSoruce.token
    };
    const { data } = await axios(options);
    // cancelTokenSoruce = null;
    return data;
};

export const useGetDataFromServer = (url: string, payload: Payload, enabled: boolean) => {
    return useQuery([url, payload], () => {
        return getDataFromServer(url, payload);
    }, { enabled });
};

