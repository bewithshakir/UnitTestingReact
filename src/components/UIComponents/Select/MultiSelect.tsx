import { FormikErrors } from 'formik';
import { Fragment } from 'react';
import Select, { components, DropdownIndicatorProps, OptionProps} from 'react-select';
import { FormHelperText, InputLabel, FormControl } from '@mui/material';
import './SingleSelect.scss';
import { ArrowDown, CheckedCheckboxIcon as Check, UncheckedCheckboxIcon as UnCheck } from '../../../assets/icons';


type item = {
    label: string,
    value: string | number
}
interface props {
    id?: string;
    name?: string;
    label: string;
    placeholder?: string;
    value?: Array<item>;
    items: Array<item>;

    required?: boolean;
    error?: boolean;
    helperText?: string | FormikErrors<item> | undefined;
    loadingMessage?: string;
    noOptionsMessage?: string;

    description?: string;
    onChange: (...args: any[]) => void;
    onBlur?: (...args: any[]) => void;

    disabled?:boolean
}

export default function MultiSelect(props: props) {
    

    const MultiValue = (props:any) => {
        const {index, getValue } = props;
        const value = getValue();
        const length = value.length;
        const abLength = length - 2 ;

        if (index > 0) return null;

        return(length > 0) ?
                <div className="option-label">
                {length === 1 && (value[0].label)}
                {length === 2 && ([value[0].label, 'and', value[1].label].join(' '))}
                {length > 2 && ([`${value[0].label},`, value[1].label, 'and', abLength, 'more'].join(' '))}
                </div> : <components.MultiValue {...props} />
        ;
    };


    const DropdownIndicator = (props: DropdownIndicatorProps<any>) => {
        return (
            <components.DropdownIndicator {...props}>
                <ArrowDown />
            </components.DropdownIndicator>
        );
    };

    const Option = (props: OptionProps<any>) => {
        return (
            <components.Option {...props}>
                <Fragment>
                    <span className='box'>
                        {props.isSelected ? <Check /> : <UnCheck />}
                        <label>{props.label}</label>
                    </span>
                </Fragment>
            </components.Option>
        );
    };
    const handleChange = (e: any) => {
        const { name, onChange } = props;
        onChange(name, e);
    };

    return (
        <Fragment>
            <FormControl className='select' >
                <InputLabel shrink htmlFor={`${props.id}-label`} style={{ color: 'var(--Darkgray)' }} aria-labelledby={props.label} aria-required={props.required}>
                    <b>{props.label.toUpperCase()}{props.required && props.label && (<span className='super' >* </span>)}</b >
                </InputLabel>
                <Select
                    id={props.id}
                    name={props.name}
                    placeholder={props.placeholder}
                    className={props.error ? 'multi-select-container error' : 'multi-select-container'}
                    classNamePrefix='multi-select'
                    value={props.value}
                    options={props.items}
                    onChange={handleChange}
                    onBlur={props.onBlur}
                    components={{ ClearIndicator: () => null, IndicatorSeparator: () => null, Option, DropdownIndicator, MultiValue}}
                    isSearchable={false}
                    hideSelectedOptions={false}
                    closeMenuOnSelect={false}
                    isMulti={true}
                    noOptionsMessage={() => props.noOptionsMessage || null}
                    loadingMessage={() => props.loadingMessage || null}
                    isDisabled={props.disabled}
                />
                {props.helperText && (
                    <FormHelperText
                        className='select-helper'
                        id={props.description}
                        error={props.error}>
                        {props.helperText}
                    </FormHelperText>)}
            </FormControl>
        </Fragment>
    );


}

MultiSelect.defaultProps = {
    required: false,
    id: "select-label",
    error: false,
    placeholder: 'Select',
    noOptionsMessage: 'No Options found',
    loadingMessage: 'Loading...'
};
