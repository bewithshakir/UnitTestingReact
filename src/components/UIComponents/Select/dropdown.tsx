import { useState, useEffect, Fragment } from 'react';
import { DropSearchIcon as SearchIcon, Check, ArrowUp, ArrowDown, UnCheck } from '../../../assets/icons';
import './dropdown.scss';

type item = {
  label: string,
  value: string | number
}
interface props {
  name?: string;
  label: string;
  search?: boolean;
  placeholder?: string;
  multiple?: boolean;
  value?: Array<item>;
  width?: number | string;
  items: Array<item>;
  onChange: (...args: any[]) => void;
}

export const Divider = () => <div className='divider'></div>;


export default function Select(props: props) {
  const [selectedValues, setSelectedValues] = useState<any>([]);
  const [selectProps, setSelectedProps] = useState({ focusedValue: -1, isFocused: false, isOpen: false });
  const [searchTerm, setSearchTerm] = useState('');
  const [searchResults, setSearchResults] = useState<any[]>([]);

  useEffect(() => {
    const { value } = props;
    value && value.length > 0 && setSelectedValues([...value])
  }, [])

  const onFocus = () => setSelectedProps(x => ({ ...x, isFocused: true }));
  const onClick = () => {
    const { isOpen } = selectProps;
    setSelectedProps(x => ({ ...x, isOpen: !isOpen }));
  }
  const onHoverOption = (e: any) => {
    const { items } = props;
    const { value } = e.currentTarget.dataset;
    const index = items.findIndex(i => i.value === value);
    setSelectedProps(x => ({ ...x, focusedValue: index }));
  }

  const onChange = () => {
    const { name, onChange } = props;
    let obj = {
      target: {
        name: name,
        value: selectedValues
      }
    }
    onChange(obj);
  }

  useEffect(() => onChange(), [selectedValues.length]);

  const onClickOption = (e: any) => {
    const { multiple } = props;
    const { value } = e.currentTarget.dataset;
    const index = searchResults.findIndex(i => i.value === value);
    const index1 = selectedValues.findIndex((s: any) => s.value === value);
    if (multiple) {
      if (index1 === -1) {
        selectedValues.push(searchResults[index])
      } else {
        selectedValues.splice(index1, 1)
      }
      setSelectedValues([...selectedValues]);
    } else {
      setSelectedValues([searchResults[index]]);
      setSelectedProps(x => ({ ...x, isOpen: false }));
    }
  }

  const onBlur = () => {
    const { items, multiple } = props;
    if (multiple) {
      setSelectedProps(x => ({ ...x, isOpen: false, isFocused: false, focusedValue: -1 }));
    } else {
      const value = selectedValues[0]
      let focusedValue = -1
      if (value) {
        focusedValue = items.findIndex(option => option.value === value)
      }
      setSelectedProps(x => ({ ...x, isOpen: false, isFocused: false, focusedValue: focusedValue }));
    }
  }

  const onKeyDown = (e: React.KeyboardEvent) => {
    const { items, multiple } = props
    const { isOpen } = selectProps;
    let { focusedValue } = selectProps;
    switch (e.key) {     
      case 'Esc':
        if (isOpen) {
          e.preventDefault()
          setSelectedProps(x => ({ ...x, isOpen: false }));
        }
        break
      case 'Tab':
        setSelectedProps(x => ({ ...x, isOpen: !isOpen }));
        break
      case 'Enter':
        setSelectedProps(x => ({ ...x, isOpen: true }));
        break
      case 'ArrowDown':
        e.preventDefault()
        if (focusedValue < items.length - 1) {
          focusedValue++

          if (multiple) {
            setSelectedProps(x => ({ ...x, focusedValue: focusedValue }));
          } else {
            setSelectedValues([items[focusedValue].value]);
            setSelectedProps(x => ({ ...x, focusedValue: focusedValue }));
          }
        }

        break
      case 'ArrowUp':
        e.preventDefault()
        if (focusedValue > 0) {
          focusedValue--

          if (multiple) {
            setSelectedProps(x => ({ ...x, focusedValue: focusedValue }));
          } else {
            setSelectedValues([items[focusedValue].value]);
            setSelectedProps(x => ({ ...x, focusedValue: focusedValue }));
          }
        }
        break
      default:
        e.preventDefault()
        if (isOpen) {
          if (multiple) {
            if (focusedValue !== -1) {
              const value = items[focusedValue].value
              const index = selectedValues.indexOf(value)

              if (index === -1) {
                selectedValues.push(value)
              } else {
                selectedValues.splice(index, 1)
              }
              setSelectedValues([...selectedValues]);
            }
          }
        } else {
          setSelectedProps(x => ({ ...x, isOpen: true }));
        }
        break
    }
  }


  const searchHandler = (searchTerm: string) => {
    const { items } = props;
    if (searchTerm !== '') {
      const newOptionList = items.filter((items) => {
        return Object.values(items)
          .join('')
          .toLowerCase()
          .includes(searchTerm.toLowerCase());
      });
      setSearchResults(newOptionList);
    } else {
      setSearchResults(items);
    }
  }

  const getSearchTerm = (e: any) => { setSearchTerm(e.target.value); }

  useEffect(() => searchHandler(searchTerm), [searchTerm]);

  const renderSearch = () => {
    const { search } = props;
    return (
      <Fragment>
        {search && (
          <div className='ui search'>
            <div className='ui icon input'>
              <SearchIcon />
              <input
                type='text'
                placeholder='Search'
                className='prompt'
                value={searchTerm}
                onChange={getSearchTerm}
                style={{ paddingLeft: 12 }}
              />
            </div>
          </div>
        )}
      </Fragment>
    )
  }

  const renderValues = () => {
    const { placeholder, multiple } = props;
    const length = selectedValues.length;
    const ablength = length - 1;

    if (selectedValues.length === 0) {
      return (
        <div className="placeholder">
          {placeholder}
        </div>
      )
    }

    if (multiple) {
      return (
        <div className="value">
          {length === 1 && (selectedValues[0].label)}
          {length > 1 && ([selectedValues[0].label, 'and', ablength, 'more'].join(' '))}
        </div>
      )
    }

    return (
      <div className="value">
        {selectedValues[0].label}
      </div>
    )
  }

  const renderOption = (option: any, index: number) => {
    const { multiple } = props;
    const { focusedValue } = selectProps;

    const { value, label } = option;

    const selected = selectedValues.includes(option);

    let className = "option"
    if (selected) className += " selected"
    if (index === focusedValue) className += " focused"

    return (
      <div
        key={value}
        data-value={value}
        className={className}
        onMouseOver={onHoverOption}
        id={props.name}
        onClick={onClickOption}
      >
        {multiple ?
          <span className="checkbox">
            {selected ? <Check /> : <UnCheck />}
          </span> :
          null
        }
        {label}
      </div>
    )
  }

  const renderOptions = () => {
    const { search, multiple } = props;
    const { isOpen } = selectProps;
    const length = selectedValues.length;

    if (!isOpen) {
      return null
    }

    return (
      <div className="options">
        {search && (<div className="opt-search">
          {renderSearch()}
          <Divider />
        </div>)}
        <div className="opt-items">
          {multiple && (<div className="selected-items">
            {length > 0 && (selectedValues.map(renderOption))}
            {length > 0 && (<Divider />)}
          </div>)}
          <div className='items'>
            {searchResults.map(renderOption)}
          </div>
        </div>
      </div>
    )
  }


  return (
    <div
      className="select-dropdown"
      style={{ width: props.width }}
      onFocus={onFocus}
      onBlur={onBlur}
      onKeyDown={onKeyDown}
      tabIndex={0}
    >
      <label className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink label"><b>{props.label.toUpperCase()}</b></label>
      <div className={selectProps.isOpen ? "selection open" : "selection"} onClick={onClick}>
        {renderValues()}
        <span className="arrow">
          {selectProps.isOpen ? <ArrowUp /> : <ArrowDown />}
        </span>
      </div>
      {renderOptions()}
    </div>
  )

}

Select.defaultProps = {
  search: false,
  id: "select-label",
  multiple: false,
  width: "100%",
}