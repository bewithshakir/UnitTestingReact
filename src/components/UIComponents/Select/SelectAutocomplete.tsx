import { FormikErrors } from 'formik';
import { Fragment } from 'react';
import Select from 'react-select';
import { FormHelperText, InputLabel, FormControl } from '@mui/material';
import './SingleSelect.scss';

type item = {
    label: string,
    value: string | number
}

interface props {
    id?: string;
    name?: string;
    label?: string;
    placeholder?: string;
    value: item | string | null;
    items?: Array<item>;

    required?: boolean;
    error?: boolean;
    helperText?: string | FormikErrors<item> | undefined;

    isDisabled?: boolean;

    description?: string;
    onChange: (...args: any[]) => void;
    onBlur?: (...args: any[]) => void;
    onInputChange: (...args: any[]) => void;
    disabled?: boolean;
    components?: any;
}

export default function SelectAutoComplete(props: props) {

    const handleInputChange =(keyword: string) => {
        props.onInputChange(keyword);
    };

    const handleChange = (e: any) => {
        const { name, onChange } = props;
        onChange(name, e);
    };

    return (
        <Fragment>
            <FormControl className='select' >
                {props.label && (<InputLabel shrink htmlFor={`${props.id}-label`} style={{ color: 'var(--Darkgray)' }} aria-labelledby={props.label} aria-required={props.required}>
                    <b>{props.label.toUpperCase()}{props.required && props.label && (<span className='super' >* </span>)}</b >
                </InputLabel>)}
                <Select
                    id={props.id}
                    name={props.name}
                    placeholder={props.placeholder}
                    className={props.error ? 'react-select-container error' : 'react-select-container'}
                    classNamePrefix='react-select'
                    value={props.value}
                    options={props.items}
                    onChange={handleChange}
                    onInputChange={handleInputChange}
                    onBlur={props.onBlur}
                    components={{ IndicatorSeparator: () => null, DropdownIndicator: () => null}}
                    isSearchable={true}
                    isDisabled={props.disabled}
                    isMulti={false}
                />
                {props.helperText && (
                    <FormHelperText
                        className='select-helper'
                        id={props.description}
                        error={props.error}>
                        {props.helperText}
                    </FormHelperText>)}
            </FormControl>
        </Fragment>
    );


}

SelectAutoComplete.defaultProps = {
    required: false,
    id: "select-label",
    error: false,
    placeholder: 'Search Location',
};
