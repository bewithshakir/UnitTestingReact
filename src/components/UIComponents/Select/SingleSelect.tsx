import { FormikErrors } from 'formik';
import { Fragment } from 'react';
import Select, { components, DropdownIndicatorProps, OptionProps } from 'react-select';
import { FormHelperText, InputLabel, FormControl, Icon, Box, Typography } from '@mui/material';
import './SingleSelect.scss';
import { ArrowDown } from '../../../assets/icons';

const { SingleValue } = components;
const optionIconsSX = { width: "20px", height: "20px" };

type item = {
    label: string,
    value: string | number,
    icon?: (props: React.SVGProps<SVGSVGElement>) => JSX.Element
}
interface props {
    id?: string;
    name?: string;
    label?: string;
    placeholder?: string;
    value?: item | Array<item>;
    items: Array<item> | [];

    required?: boolean;
    error?: boolean;
    helperText?: string | FormikErrors<item> | undefined;

    isDisabled?: boolean;

    description?: string;
    onChange: (...args: any[]) => void;
    onBlur?: (...args: any[]) => void;
    disabled?: boolean;
    components?: any;
    isLoading?: boolean;
    noOptionsMessage?: () => string
}

export default function SingleSelect (props: props) {

    const DropdownIndicator = (props: DropdownIndicatorProps<item>) => {
        return (
            <components.DropdownIndicator {...props}>
                <ArrowDown />
            </components.DropdownIndicator>
        );
    };

    const handleChange = (e: any) => {
        const { name, onChange } = props;
        onChange(name, e);
    };

    const Option = (props: OptionProps<any>) => {
        return (
            <components.Option {...props}>
                <Box display="flex" alignItems="center" justifyContent="start">
                    {props.data.icon ? <Icon sx={optionIconsSX} component={props.data.icon} /> : null}
                    {props.data.label ? <Typography variant="h4" pl={props.data.icon ? 1 : 0}>{props.data.label}</Typography> : null}
                </Box>
            </components.Option>
        );
    };

    const SingleValueFrag = (props: any) => {
        return (
            <SingleValue {...props}>
                <Box display="flex" alignItems="center" justifyContent="start">
                    {props.data.icon ? <Icon sx={optionIconsSX} component={props.data.icon} /> : null}
                    {props.data.label ? <Typography variant="h4" pl={props.data.icon ? 1 : 0}>{props.data.label}</Typography> : null}
                </Box>
            </SingleValue>
        );
    };

    const checkIsEmptyOption = (value: item | Array<item> | undefined) => {
        if (value && !Array.isArray(value)) {
            if (value?.label.toString().length && value?.value.toString().length) {
                return value;
            } else {
                return null;
            }
        } else {
            return value;
        }
    };

    return (
        <Fragment>
            <FormControl className='select' >
                {props.label && (<InputLabel shrink htmlFor={`${props.id}`} style={{ color: 'var(--Darkgray)' }} aria-labelledby={props.label} aria-required={props.required}>
                    <b>{props.label.toUpperCase()}{props.required && props.label && (<span className='super' >* </span>)}</b >
                </InputLabel>)}
                <Select
                    inputId={props.id}
                    name={props.name}
                    placeholder={props.placeholder}
                    className={props.error ? 'react-select-container error' : 'react-select-container'}
                    classNamePrefix='react-select'
                    value={checkIsEmptyOption(props.value)}
                    options={props.items}
                    onChange={handleChange}
                    onBlur={props.onBlur}
                    components={{ IndicatorSeparator: () => null, DropdownIndicator, ...props.components, Option, SingleValue: SingleValueFrag }}
                    isSearchable={false}
                    isDisabled={props.isDisabled}
                    isLoading={props?.isLoading}
                    noOptionsMessage={props.noOptionsMessage}
                />
                {props.helperText && (
                    <FormHelperText
                        className='select-helper'
                        id={props.description}
                        error={props.error}>
                        {props.helperText}
                    </FormHelperText>)}
            </FormControl>
        </Fragment>
    );


}

SingleSelect.defaultProps = {
    required: false,
    id: "select-label",
    error: false,
    placeholder: 'Select',
};
