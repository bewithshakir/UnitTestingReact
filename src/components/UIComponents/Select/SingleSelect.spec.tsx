/* eslint-disable no-console */
import Select from './SingleSelect';
// import { YellowFuelIcon } from '../../../assets/icons';
import { shallow, mount } from 'enzyme';
import { getProductIcon } from '../../../utils/helperFunctions';

describe('renders the Default Select component', () => {
    const component = shallow(<Select label='Input' items={[{ label: 'Label', value: 'Value', icon: getProductIcon('Yellow') }]} onChange={() => jest.fn()} />);

    it('snapshot test for Default Select component', () => {
        expect(component).toMatchSnapshot();
    });

});

describe('renders the SingleSelect component with values', () => {
    const component = shallow(<Select label='Input' value={{ label: 'Label1', value: 'Value1', icon: getProductIcon('Yellow') }} items={[{ label: 'Label1', value: 'Value1', icon: getProductIcon('Yellow') }, { label: 'Label2', value: 'Value2', icon: getProductIcon('Yellow') }, { label: 'Label3', value: 'Value3', icon: getProductIcon('Yellow') }]} onChange={() => jest.fn()} />);

    it('Snapshot test for the SingleSelect component ', () => {
        expect(component).toMatchSnapshot();
    });
});


describe('renders the SingleSelect component with error ', () => {
    const component = shallow(<Select label='Input' error={true} value={{ label: 'Label1', value: 'Value1', icon: getProductIcon('Yellow') }} items={[{ label: 'Label1', value: 'Value1', icon: getProductIcon('Yellow') }, { label: 'Label2', value: 'Value2', icon: getProductIcon('Yellow') }, { label: 'Label3', value: 'Value3', icon: getProductIcon('Yellow') }]} onChange={() => jest.fn()} />);

    it('snapshot test for the SingleSelect component ', () => {
        expect(component).toMatchSnapshot();
    });
});

describe('renders the SingleSelect component with required prop', () => {
    const component = shallow(<Select label='Input' required value={{ label: 'Label1', value: 'Value1', icon: getProductIcon('Yellow') }} items={[{ label: 'Label1', value: 'Value1', icon: getProductIcon('Yellow') }, { label: 'Label2', value: 'Value2', icon: getProductIcon('Yellow') }, { label: 'Label3', value: 'Value3', icon: getProductIcon('Yellow') }]} onChange={() => jest.fn()} />);

    it('snapshot test for the SingleSelect component ', () => {
        expect(component).toMatchSnapshot();
    });
});

describe('renders the SingleSelect component with helper text', () => {
    const component = shallow(<Select label='Input' required helperText={'sample error'} value={{ label: 'Label1', value: 'Value1', icon: getProductIcon('Yellow') }} items={[{ label: 'Label1', value: 'Value1', icon: getProductIcon('Yellow') }, { label: 'Label2', value: 'Value2', icon: getProductIcon('Yellow') }, { label: 'Label3', value: 'Value3', icon: getProductIcon('Yellow') }]} onChange={() => jest.fn()} />);

    it('snapshot test for the SingleSelect component ', () => {
        expect(component).toMatchSnapshot();
    });
});

describe('renders the SingleSelect component on change', () => {
    it('Should change value when onChange was called', () => {
        const props = {
            label: 'Select values',
            onChange: jest.fn(),
            id: 'single-select-on-change-test',
            items: [{ label: 'Label1', value: 'Value1', icon: getProductIcon('Yellow') }, { label: 'Label2', value: 'Value2', icon: getProductIcon('Yellow') }, { label: 'Label3', value: 'Value3', icon: getProductIcon('Yellow') }]
        };
        const wrapper = mount(<Select {...props} />);
        const event = { label: 'Label1', value: 'Value1', icon: getProductIcon('Yellow') };
        wrapper.find('Select').simulate('change', event);
        expect(wrapper).toMatchSnapshot();
    });
});


describe('renders the SingleSelect component with correct no of options', () => {
    it('Should render correct list of options', () => {
        const options = [{ label: 'Label1', value: 'Value1', icon: getProductIcon('Yellow') }, { label: 'Label2', value: 'Value2', icon: getProductIcon('Yellow') }];
        const props = {
            label: 'Select values',
            onChange: jest.fn(),
            id: 'single-select',
            items: options
        };
        const wrapper = mount(<Select {...props} />);
        wrapper.find('div.react-select__indicator').simulate('mouseDown', { button: 0 });
        expect(wrapper.find('div.react-select__option').length).toEqual(options.length);
    });
});

describe('renders the SingleSelect component with any option selected', () => {
    it('Should render correct text in value', () => {
        const options = [{ label: 'Label1', value: 'Value1', icon: getProductIcon('Yellow') }, { label: 'Label2', value: 'Value2', icon: getProductIcon('Yellow') }, { label: 'Label3', value: 'Value3', icon: getProductIcon('Yellow') }];
        const randomNumber = Math.floor(Math.random() * ((options.length - 1) - 0 + 1)) + 0;
        const props = {
            label: 'Select values',
            onChange: jest.fn(),
            id: 'single-select-options-check',
            items: options
        };
        const wrapper = mount(<Select {...props} />);
        wrapper.find('div.react-select__dropdown-indicator').simulate('mouseDown', { button: 0 });
        wrapper.find('div.react-select__option').at(randomNumber).simulate('click', null);
        expect(wrapper.find('h4').text()).toEqual(options[randomNumber].label);
    });
});
