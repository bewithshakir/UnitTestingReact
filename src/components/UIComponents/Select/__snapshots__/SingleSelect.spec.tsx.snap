// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the Default Select component snapshot test for Default Select component 1`] = `
<Fragment>
  <ForwardRef(FormControl)
    className="select"
  >
    <ForwardRef(InputLabel)
      aria-labelledby="Input"
      aria-required={false}
      htmlFor="input-label"
      shrink={true}
      style={
        Object {
          "color": "var(--Darkgray)",
        }
      }
    >
      <b>
        INPUT
      </b>
    </ForwardRef(InputLabel)>
    <ForwardRef
      className="react-select-container"
      classNamePrefix="react-select"
      components={
        Object {
          "DropdownIndicator": [Function],
          "IndicatorSeparator": [Function],
          "Option": [Function],
          "SingleValue": [Function],
        }
      }
      id="input"
      isSearchable={false}
      onChange={[Function]}
      onKeyDown={[Function]}
      options={
        Array [
          Object {
            "icon": <YellowFuelIcon />,
            "label": "Label",
            "value": "Value",
          },
        ]
      }
      placeholder="Select"
    />
  </ForwardRef(FormControl)>
</Fragment>
`;
